src/typing/env_js.ml

```diff
@@ -293,49 +342,93 @@ let already_bound_error =
    (since multiple declarations - sometimes but not always erroneous -
    may appear in an AST)
  *)
-let bind_entry cx name entry =
-  let scope = peek_scope () in
-  match Scope.get_entry name scope with
 
+let bind_entry cx name entry reason =
+  (* lex scopes can only hold let/const bindings
+   * var scopes can hold all bindings
+   * var bindings can't shadow lex bindings
+   * type entries are hoisted to var scope *)
+  let rec find_scope = function
+    | [] -> assert_false "empty scope list"
+    | scope::scopes -> Scope.(
+      match scope.kind, entry with
+      | LexScope, Entry.Value { Entry.kind = Entry.Let _; _ }
+      | LexScope, Entry.Value { Entry.kind = Entry.Const; _ }
+      | VarScope _, _ ->
+        Some scope
+      | LexScope, Entry.Type _ ->
+        find_scope scopes
+      | LexScope, Entry.Value { Entry.kind = Entry.Var; _ } ->
+        match get_entry name scope with
+        (* nested scope var shadows a lex-scoped lexical *)
+        | Some prev
+          when is_lex scope && Entry.is_lex prev ->
+            already_bound_error cx name prev reason;
+            None
+        (* keep searching for the var's scope *)
+        | _ -> find_scope scopes
```

Comment:
This pattern will fire for the `Some _` case when the side condition is false, causing us to look past existing entries. Don't think that's right, is it?